src/librbd/io/ObjectRequest.cc

```diff
@@ -521,11 +522,26 @@ void AbstractObjectWriteRequest<I>::handle_write_object(int r) {
   ldout(image_ctx->cct, 20) << "r=" << r << dendl;
 
   r = filter_write_result(r);
-  if (r == -ENOENT || (r == -ERANGE && !image_ctx->migration_info.empty())) {
-    if (m_copyup_enabled) {
+  if (r == -ENOENT && m_copyup_enabled) {
+    copyup();
+    return;
+  } else if (r == -ERANGE && m_guarding_migration_write) {
+    if (!image_ctx->migration_info.empty()) {
       copyup();
-      return;
+    } else {
+      ldout(image_ctx->cct, 10) << "migration parent gone, restart io" << dendl;
+      m_guarding_migration_write = false;
+      {
```

Comment:
Nit: while perhaps not 100% optimal, I wonder if this can just be removed for clarity (i.e. keep ```m_copyup_enabled == true```) and just re-issue the write. If the parent no longer exists, it will hit ```-ENOENT``` and the copy-up state machine will be invoked (which will check for the existence of the parent.

Reply:
@dillaman We pass `m_parent_extents` to the copyup request, which is calculated for the migration parent, if we don't recalculate it. And I don't see `CopyupRequest::send` checks from the existence of the parent.

Reply:
@trociny Perhaps then just abstract the ```compute_parent_extents``` and ```m_copyup_enabled``` logic to a new function and re-use it here and in the constructor.